02_bubble/alfredxiao/main.go

```diff
@@ -0,0 +1,51 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func makeCopy(ns []int) []int {
+	return append([]int{}, ns...)
+}
+
+func bubble(s []int) []int {
+	copy := makeCopy(s)
+	n := len(copy)
+	for i := n - 1; i > 0; i-- {
+		for j := 0; j < i; j++ {
+			maybeSwap(copy, j, j+1)
+		}
+	}
+	return copy
+}
+
+func maybeSwap(ns []int, i, j int) bool {
+	if ns[i] > ns[j] {
+		ns[i], ns[j] = ns[j], ns[i]
+		return true
+	}
+	return false
+}
+
+func insertion(ns []int) []int {
```

Comment:
you forgot to copy input slice. `s := makeCopy(ns)`

Reply:
good catch